### ISSUE TITLE
Unit-test: fatal error in "GoodSubscription": signal: illegal opcode; address of failing instruction: 0x00409d78

### ISSUE THREAD

**Issue by anonimal** [monero-project/kovri#699]
>Running 140 test cases...
[2017-08-29 21:13:58.836303] [0xb6f46000] [debug]   AddressBook: validating subscription
unknown location(0): fatal error in "GoodSubscription": signal: illegal opcode; address of failing instruction: 0x00409d78
/home/buildbot/slave/kovri-all-ubuntu-arm7/build/tests/unit_tests/client/address_book/impl.cc(91): last checkpoint

https://build.getmonero.org/builders/kovri-all-ubuntu-arm7/builds/441/steps/test/logs/stdio

I'm 99.99% sure that this is an upstream cryptopp issue (slap me if I'm wrong). Appears to only affects *this* build machine (ARMv7 Processor rev 2 (v7l)) and not our other armv7l machine (ARMv7 Processor rev 4 (v7l)) though the rev4 machine had built the tests with clang and not gcc (still waiting for its gcc build to finish).

It's possible that calculating the SHA256 digest when verifying the base64 ident in address book could trip an illegal opcode if ARM features weren't properly detected (TBD). I've noticed several cryptopp ARM-related issues in their current master (as we can see now when https://github.com/monero-project/kovri/pull/693 was introduced). I'll open new tickets with them a.s.a.p.

Note: the armv8 machine is still down and has yet to be tested with #693.

---
**By submitting this issue, I confirm the following:**

- I have read and understood the contributor guide in [kovri-docs](https://github.com/monero-project/kovri-docs).
- I have checked that the issue I am reporting can be replicated or that the feature I am suggesting is not present.
- I have checked opened or recently closed [pull requests](https://github.com/monero-project/kovri/pulls) for existing solutions/implementations to my issue/suggestion.
---

**Comment by noloader** [comment]
Hi @anonimal,

Sorry to jump over to another bug reporter.

> unknown location(0): fatal error in "GoodSubscription": signal: illegal opcode; address of failing instruction: 0x00409d78

I'm guessing the `unknown location(0)` is because feature detection occurs in `DetectArmFeatures`. Its a static class and it runs as a C++ initializer. Also see [`cpu.cpp` : 555](https://github.com/weidai11/cryptopp/blob/master/cpu.cpp#L555). That probably explains the weird location report:

```
void DetectArmFeatures()
{
	g_hasNEON  = CPU_QueryNEON() || CPU_ProbeNEON();
	...
	g_ArmDetectionDone = true;
}

struct InitializeCpu
{
	InitializeCpu()
	{
#if CRYPTOPP_BOOL_X86 || CRYPTOPP_BOOL_X32 || CRYPTOPP_BOOL_X64
		CryptoPP::DetectX86Features();
#elif CRYPTOPP_BOOL_ARM32 || CRYPTOPP_BOOL_ARM64
		CryptoPP::DetectArmFeatures();
#endif
	}
};

```

> `signal: illegal opcode; address of failing instruction: 0x00409d78`

We should have a signal handler installed to catch a `SIGILL`. I'm wondering if BuildBot is prematurely killing the process, or if we screwed up the probe.

In the bigger picture, this is what we have in place. First, we perform the feature probes/queries once and retain the result ([`cpu.cpp`](https://github.com/weidai11/cryptopp/blob/master/cpu.cpp)):

```
void DetectArmFeatures()
{
	g_hasNEON  = CPU_QueryNEON() || CPU_ProbeNEON();
	...

	g_ArmDetectionDone = true;
}
```

Second, the `CPU_QueryXXX` tries to find the feature with `getauxvalue` (and friends) ([`cpu.cpp`](https://github.com/weidai11/cryptopp/blob/master/cpu.cpp)):

```
inline bool CPU_QueryNEON()
{
#if defined(__ANDROID__) && (defined(__aarch32__) || defined(__aarch64__))
	if (android_getCpuFeatures() & ANDROID_CPU_ARM64_FEATURE_ASIMD)
		return true;
#elif defined(__ANDROID__) && defined(__arm__)
	if (android_getCpuFeatures() & ANDROID_CPU_ARM_FEATURE_NEON)
		return true;
#elif defined(__linux__) && defined(__aarch64__)
	if (getauxval(AT_HWCAP) & HWCAP_ASIMD)
		return true;
#elif defined(__linux__) && defined(__aarch32__)
	if (getauxval(AT_HWCAP2) & HWCAP2_ASIMD)
		return true;
#elif defined(__linux__) && defined(__arm__)
	if (getauxval(AT_HWCAP) & HWCAP_ARM_NEON)
		return true;
#elif defined(__APPLE__) && defined(__aarch64__)
	// Core feature set for Aarch32 and Aarch64.
	return true;
#endif
	return false;
}
```

Third, if the *Query* fails, we move onto a *Probe* that uses `SIGILL` probes ([`neon-simd.cpp`](https://github.com/weidai11/cryptopp/blob/master/neon-simd.cpp)):

```
bool CPU_ProbeNEON()
{
    ...
    volatile bool result = true;

    volatile SigHandler oldHandler = signal(SIGILL, SigIllHandler);
    if (oldHandler == SIG_ERR)
        return false;

    volatile sigset_t oldMask;
    if (sigprocmask(0, NULLPTR, (sigset_t*)&oldMask))
        return false;

    if (setjmp(s_jmpSIGILL))
        result = false;
    else
    {
        uint32_t v1[4] = {1,1,1,1};
        uint32x4_t x1 = vld1q_u32(v1);
        uint64_t v2[2] = {1,1};
        uint64x2_t x2 = vld1q_u64(v2);
        ....

        // Hack... GCC optimizes away the code and returns true
        result = !!(vgetq_lane_u32(x3,0) | vgetq_lane_u64(x4,1));
    }

    sigprocmask(SIG_SETMASK, (sigset_t*)&oldMask, NULLPTR);
    signal(SIGILL, oldHandler);
    return result;
}
```

Maybe we need a `CRYPTOPP_NO_CPU_FEATURE_PROBES` to disable the probes altogether (but not the queries).

**Comment by anonimal** [comment]
Hi @noloader, thanks for the info.

>I'm wondering if BuildBot is prematurely killing the process, or if we screwed up the probe.

This is reproducible without BuildBot.

>Maybe we need a CRYPTOPP_NO_CPU_FEATURE_PROBES to disable the probes altogether (but not the queries).

@noloader, I'll open a cryptopp issue/feature request.

**Comment by anonimal** [comment]
@noloader https://github.com/weidai11/cryptopp/issues/511 does not resolve this issue and the issue now appears to extend to x86-64:

```
Linux 4.12.13-1-ARCH #1 SMP PREEMPT Fri Sep 15 06:36:43 UTC 2017 x86_64 GNU/Linux
gcc (GCC) 7.2.0
```

With our build patched to build flags `-march=native -DCRYPTOPP_NO_CPU_FEATURE_PROBES=1`:

```
[2017-09-20 01:57:53.400805] [0x00007f5cdf68afc0] [debug]   AddressBook: validating subscription
unknown location(0): fatal error: in "AddressBook/GoodSubscription": signal: SIGSEGV, si_code: 0 (memory access violation at address: 0x00000080)
/home/anonimal/kovri/tests/unit_tests/client/address_book/impl.cc(91): last checkpoint

*** 1 failure is detected in the test module "Master Test Suite"
```

I'm waiting for the ARMv7 build box to finish building. Will report back then.

**Comment by anonimal** [comment]
@noloader note, so far, on x86-64, cryptest passes with `-DCRYPTOPP_NO_CPU_FEATURE_PROBES=1` but I noticed that `-march=native` is only applied to the c++03 variant. We are building with the c++14 standard. I added another test to cryptest for c++14 but the test still passes. 😕

**Comment by noloader** [comment]
> ```
> [2017-09-20 01:57:53.400805] [0x00007f5cdf68afc0] [debug]   AddressBook: validating subscription
> unknown location(0): fatal error: in "AddressBook/GoodSubscription": signal: SIGSEGV, si_code: 0 (memory access violation at address: 0x00000080)
> /home/anonimal/kovri/tests/unit_tests/client/address_book/impl.cc(91): last checkpoint
> ```

The `memory access violation at address: 0x00000080` looks a little suspicious. It could be a vtable dereference for an object not yet constructed. I.e., the [Static Initialization Order Fiasco](https://www.cryptopp.com/wiki/Static_Initialization_Order_Fiasco) problem.

Do you have any global C++ objects in play? If so, then the order of initialization is important. Linkers are deterministic, so if an C++ object in `B` depends on a C++ object in `A`, then you should order `a.o` before `b.o`. The linker will then call the constructors in `A` before those in `B`.

Crypto++ handles it within the library using priority attributes, like seen at [`cryptlib.cpp` : 930](https://github.com/weidai11/cryptopp/blob/master/cryptlib.cpp#L929). The attributes ensure a single object is created and always returned. For example, the same object is always returned for `DEFAULT_CHANNEL` and `Integer::One()`.

Less popular objects get created on the fly as needed with the `Singleton` pattern. Its non-counted so new objects are created as needed if the compiler is not smart enough to combine objects. This is needed because of Microsoft's broken dynamic initialization implementation in Windows 8 and below. For non-Windows C++11 and above, we use dynamic initialization.

If you think Crypto++ is the problem, then add `-DCRYPTOPP_INIT_PRIORITY=0`. The `0` priority value signals you do not want to use them.

**Comment by noloader** [comment]
` I noticed that -march=native is only applied to the c++03 variant....`

Well, I'm not sure about this.

We removed `-march=native` by default at [Issue #380, Random crashes on different computers because option -march=native is active 'by default'](https://github.com/weidai11/cryptopp/issues/380#issuecomment-323962162). It happened after we split the sources at [PR #461, Split source files to support Base Implementation + SIMD implementation](https://github.com/weidai11/cryptopp/pull/461). That is, we now use `rijndael.cpp` + `rijndael-simd.cpp`, and  `rijndael-simd.cpp` gets the flags it needs to compile the source file.

**Comment by anonimal** [comment]
>Well, I'm not sure about this.

Ah, it was probably caught by `USER_CXXFLAGS` from a previous test.

>If you think Crypto++ is the problem, then add -DCRYPTOPP_INIT_PRIORITY=0.

Well, this doesn't resolve the issue either.

>Static Initialization Order Fiasco problem.

I remember reading this a while back on the wiki. We've never had a problem nor can I think of the few non-static unrelated global objects that we do have as being problematic. In fact, nothing related changed between the opening of this issue and prior to opening this issue - with the exception of cryptopp revision. 😕 

I'll look further into our end in case something was missed.

**Comment by anonimal** [comment]
Problem fixed.

@noloader take a wild guess as to what the source of the problem was *after* your fix:

.....

drum roll, please

🥁 

....

...CMake!!!!!!

😡

**Comment by noloader** [comment]
@anonimal,

>  take a wild guess as to what the source of the problem

I'm glad to hear you got to the bottom of it.

Would you happen to know the root cause? Was it CMake's selection in flags? Maybe something else, like the order of objects?

**Comment by anonimal** [comment]
@noloader, honestly, debugging CMake was such a mess that I can't recall making heads or tails of what happened. I suspect flag selection (your patch was necessary and perhaps CMake wasn't picking it up). In hindsight, the fix in #720 makes complete sense so I have no idea what CMake's problem originally was.
